---
title: "Case-study-SARS-CoV-2" 
author: "Jordi Sevilla"
institute: "PathoGenOmics-lab"
date: last-modified
date-format: "DD-MM-YYYY"
title-block-banner: true
format: 
  html: 
    page-layout: article
    embed-resources: true
    smooth-scroll: true
    theme: cosmo
    toc: true
    toc-expand: true
    toc-location: left
    toc-title: Summary
    number-sections: true
css: config/report.styles.css
editor: visual
params:
  workflow_version: ""
  div: ""
  freyja: ""
  tree: ""
  tempest: ""
  SNV: ""
  SNV_s: ""
  evo: ""
  div_value: ""
  panel: ""
  volcano: ""
  tree_ml: ""
  fig_cor_snp: ""
  stats_lm: ""
  table: ""
  sum_nv: ""
  heat_tab: ""
output-file: report.html
---
<head>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Montserrat&display=swap" rel="stylesheet">
</head>

> Workflow: `r params$workflow_version`

```{r , echo= F,message= F, include = F}
library(jsonlite)
library(gt)
library(tidyr)
library(dplyr)
library(heatmaply)
library(readr)
# values for diversity
div_values <- fromJSON(params$div_value)


# values for temp-est

stats <- fromJSON(params$stats_lm)

correlation <- stats[["r2"]]
sub_rate <- stats[["sub_rate"]]
sub_rate <- round(sub_rate, digits = 2)
p_value_lm <- stats[["pvalue"]]

# NV counts 
nv.counts <- fromJSON(params$sum_nv)

n_SNV <- nv.counts[["SNV"]]
n_INDELS <- nv.counts[["INDELS"]]

# Summary table
table <- read.csv(params$table)

# heatmap

vcf <- read_csv(params$heat_tab) 
row.names(vcf) <- vcf$`...1`
vcf <- vcf[-1]


cor.mat <- cor(vcf)

```

## Samples summary
```{r, echo= F,message= F}
gt(table) %>%
cols_label(
  Sample = "Sample",
  Collection_Date = "Collection Date",
  Lineage = "Lineage"
) %>%
tab_style(
  style = list(
    cell_text(weight = "bold")
  ),
  location = cells_column_labels()
) %>%
cols_align(
  align = "center",
  columns = everything()
) %>% 
opt_table_font(
  font = google_font("Montserrat")
)

```
## Evidence for chronic infection 



### Lineage admixture
The most probably lineage admixture for each sample has been calculated with software [Freyja](https://github.com/andersen-lab/Freyja).

![**Figure 1.** Lineage composition of samples according to Freyja. Samples are ordered by collection date.](`r params$freyja`)

### Nucleotide diversity comparison

Nucleotide diversity is has been calculated for $`r div_values[["boot.reps"]]`$ bootstrap replicates of $`r div_values[["sample.size"]]`$ samples in the context data. The distribution for the nuclotide diversity is assumed to `r div_values[["norm.text"]]` be normal with a p-value = $`r div_values[["normal.pvalue"]]`$ (*Shapiro-test*).

The nucleotide diversity value for the studied samples is $`r div_values[["diversity"]]`$. Assuming the context samples to be patient-independent, the `r div_values[["type.test"]]` p-value for a reduced nucleotide diversity in the studied samples is $`r div_values[["p.value"]]`$.

![**Figure 2.** Distribution of nucleotide diversity pi for 1000 randomly selected groups from context sequences. Red line indicates the pi value for the studied samples.](`r params$div`)

### Phylogeny and temporal signal

A maximum likelihood tree has been adjusted with the context sequences:

![**Figure 3.** Maximum-likelihood tree whith studied and context sequence. Studied sequence are represented in red.](`r params$tree_ml`)

In addition, phylogeny has been reconstructed using pairwise distances between target samples taking into account allele frequencies detected in sequencing data.

![**Figure 4.** Neighbour-joining tree for studied samples.](`r params$tree`)

With the neighbur-joining tree, partistic distances to root has been correlated with time obtaining a $R^2$ of **`r correlation`** with a p-value of $`r p_value_lm`$. The estimated substitution rate is **`r sub_rate`** substitutions per year.

![**Figure 5.** Partistic distance to root in the neighbour-joining tree against days since first sample. Red line indicates the adjusted linear regression model.](`r params$tempest`)

## Characterizing within-host evolution

### Number of polymorphic sites

Sites with minor allele frequency > 0.05  are considered as polymorphic sites.

![**Figure 6.** Number of polymorphisms over time for the studied samples.](`r params$fig_cor_snp`)

### Description for within-host nucleotide variants 

`r n_SNV` different SNV and `r n_INDELS` INDELS had been detected along the genome. In the **Figure 7a** is shown their distribution along the genome and the variation in allele frequencies in the studied samples.

::: {.panel-tabset}

## Whole genome
![**Figure 7a.** A\) Proportion of polymorphic sites along the genome calculated with 1000nt windows. B\) Nucleotide variants (NV) in the genomes of the studied samples. The shape and color indicates the type of NV. Samples are ordered by collection date.](`r params$SNV`)

## Spike sequence

![**Figure 7b.** A\) Proportion of polymorphic sites along the spike nucleotide sequence calculated with 1000nt windows. B\) Nucleotide variants (NV) in the spike sequences of the studied samples. The shape and color indicates the type of NV. Samples are ordered by collection date.](`r params$SNV_s`)
:::

### Time dependency for the within-host nucleotide variants

The correlation with time is calculated for the allele frequency of each NV.

![**Figure 8.** *Pearson* r and adjusted p-value (-log10) for the correlation with time for each SNV. Red line indicates a p-value of 0.05.](`r params$volcano`)

The significant correlated NV are swown in detail:

![**Figure 9.** Trend over time for the significantly correlated with time SNVs and for positions with more than 1 alternative allele. ](`r params$panel`)

### Correlation between alternative alleles

To see posible interactions between mutations, pairwise correlation between the alternative frequencies are calculated.

```{r, echo= F,message= F, fig.align='center'}

heatmaply_cor(
  cor.mat,
  fontsize_row = 8, 
  fontsize_col = 8,
  column_text_angle = 90)

```
![**Figure 10.** Heatmap for pairwise correlation between frquencies for alternative alleles.]()

### dn/ds over time

![**Figure 11.** Sinonymous mutations per sinonymous site **(dS)** and non sinonymous mutations mutations per non sinonymous site **(dN)** over time. The values are adjusted by the allele frequency.](`r params$evo`)



